userdata/userdata.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 Steffen Vogel.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package userdata implements a TLV parser/serializer for libnftables-compatible comments
+package userdata
+
+import (
+	"bytes"
+	"encoding/binary"
+)
+
+type Type byte
+
+const (
+	TypeComment Type = iota
+	TypeRuleID  Type = 100 // custom extension
+)
+
+func Put(udata []byte, typ Type, data []byte) []byte {
+	udata = append(udata, byte(typ), byte(len(data)))
+	udata = append(udata, data...)
+
+	return udata
+}
+
+func Get(udata []byte, styp Type) []byte {
+	for {
+		if len(udata) < 2 {
+			break
+		}
+
+		typ := Type(udata[0])
+		length := int(udata[1])
+		data := udata[2 : 2+length]
+
+		if styp == typ {
+			return data
+		}
+
+		if len(udata) < 2+length {
+			break
+		} else {
+			udata = udata[2+length:]
+		}
+	}
+
+	return nil
+}
+
+func PutInt(udata []byte, typ Type, num uint32) []byte {
```

Comment:
Should this be PutUint32, or is it somehow clear that all ints must be uint32 and there can never be any other type?

Reply:
I agree. Changed it..